Do not incorrectly add line separators for non-synthetic nodes when emitting node list

devversion opened the pull request:
As of 3c32f6e, a line separator is added between nodes if the nodes are
not synthetic and on separate lines. This is wrong and previously only
happened if the non-synthetic nodes were on different lines but had the same parent.

Fixes #44068.

devversion commented on tests/baselines/reference/objectLiteralShorthandPropertiesErrorFromNotUsingIdentifier.js:
```diff
@@ -35,7 +35,8 @@ var y = {
 var x = (_a = {
-        a: a, : .b,
+        a: a,
+        : .b,
```
It looks like these baseline changes are due to us no longer checking if previous and next nodes (when emitting the node list) are on the same line if either one does not have a parent set. We only check if nodes are on the same line if they have the same parent container. 

It looks like the logic for checking if there is a parent set has been introduced as part of the following refactoring (not sure if this was intentional; or if it's reasonable to continue doing that). https://github.com/Microsoft/TypeScript/commit/68b0323b72bf51236d8cd181ed94ba9a63f933fd.

andrewbranch commented on src/compiler/emitter.ts:
```diff
@@ -4564,15 +4564,15 @@ namespace ts {
-                            includeComments));
-                }
-                else if (!preserveSourceNewlines && !nodeIsSynthesized(previousNode) && !nodeIsSynthesized(nextNode)) {
+                else if (siblingNodePositionsAreComparable(previousNode, nextNode)) {
+                    if (preserveSourceNewlines) {
```
To avoid incurring a performance penalty during normal program emit (i.e., via `tsc`), `siblingNodePositionsAreComparable` must not be called unless `preserveSourceNewlines` is true (which only happens in the language service for determining changes for codefixes and refactors).

andrewbranch commented:
@typescript-bot perf test this — I know I made this code path cheaper than it used to be, but I think we still want to avoid it during regular emit.

andrewbranch commented:
Hm, that actually seems ok.

devversion commented:
@andrewbranch Thanks for reviewing. I've updated the logic to never call `siblingNodePositionsAreComparable` when `preserveSourceNewlines` is set to `false`. I've refactored the check slightly so that the intended behavior is more obvious. We still need the check to `getOriginalNode` but that shouldn't have a large impact on performance, but is needed for correctness.

Also I've explicitly added the fallback case for `PreferNewLine` instead of it magically being respected through the `synthesizedNodeStartsOnNewLine` else clause (which returns `true` even if the nodes are _not_ synthesized).

DanielRosenwasser commented:
@typescript-bot perf test this

DanielRosenwasser commented:
@rbuckton can you take a quick look?

@typescript-bot perf test this

DanielRosenwasser commented:
Here's what I'm thinking after chatting a bit with @andrewbranch: emit changes are hard a little bit risky. I think it's probably okay to bring this into our `master` branch, and create a cherry-pick into `release-4.3`. If we feel like this works well with nightlies for a while, we can publish a patch for 4.3. Does that sound reasonable?

DanielRosenwasser commented:
@typescript-bot cherry-pick this to release-4.3

sandersn commented:
@DanielRosenwasser this is ready to go, right? From reading your comment, it sounds like we want to get this into the nightlies as soon as possible in order to decide whether to publish a 4.3 patch.

devversion commented:
@DanielRosenwasser @sandersn Is there any update on this change? It would be great to see this change in the nightly builds, given it being a change that could be considered a regression (at least in perspective of the printer as it preserves non-existent lines between list items).

andrewbranch commented:
Let’s do it 👍
